nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
+            if bid > max_bonus_bid_from_max_stake:
+                bidders[bidder] = bid
+        return bidders
+
+    def _get_all_bonus_bidders(self, force_read: bool = False) -> dict:
+        if not force_read and self._all_bonus_bidders:
+            return self._all_bonus_bidders
+
+        bidders = self.worklock_agent.get_bidders()
+        min_bid = self.economics.worklock_min_allowed_bid
+
+        self._all_bonus_bidders = dict()
+        for bidder in bidders:
+            bid = self.worklock_agent.get_deposited_eth(bidder)
+            if bid > min_bid:
+                self._all_bonus_bidders[bidder] = bid - min_bid
+        return self._all_bonus_bidders
+
+    def _reduce_bids(self, whales: dict):
+
+        min_whale_bonus_bid = min(whales.values())
+        max_whale_bonus_bid = max(whales.values())
+
+        # first step - align at a minimum bid
+        if min_whale_bonus_bid != max_whale_bonus_bid:
+            whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+            self._all_bonus_bidders.update(whales)
+
+        bonus_eth_supply = sum(self._all_bonus_bidders.values())
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        if (min_whale_bonus_bid * bonus_worklock_supply) // bonus_eth_supply <= max_bonus_tokens:
+            raise self.BidderError(f"Internal error: At least one of bidders {whales} has allowable bid")
+
+        a = min_whale_bonus_bid * bonus_worklock_supply - max_bonus_tokens * bonus_eth_supply
+        b = bonus_worklock_supply - max_bonus_tokens * len(whales)
+        refund = -(-a//b)  # div ceil
+        min_whale_bonus_bid -= refund
+        whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+        self._all_bonus_bidders.update(whales)
+
+        return whales
+
+    def force_refund(self) -> dict:
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        whales = self.get_whales()
+        if not whales:
+            raise self.BidderError(f"All bids are allowable, force refund is not needed")
+
+        new_whales = whales
+        while new_whales:
+            whales.update(new_whales)
+            whales = self._reduce_bids(whales)
+            new_whales = self.get_whales()
+
+        receipt = self.worklock_agent.force_refund(checksum_address=self.checksum_address,
+                                                   addresses=list(whales.keys()))
+
+        if self.get_whales(force_read=True):
+            raise self.BidderError(f"Internal error: offline simulation differs from transaction results")
+        return receipt
+
+    # TODO better control: max iterations, interactive mode
+    def verify_bidding_correctness(self, gas_limit: int) -> dict:
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        if self.worklock_agent.bidders_checked():
```

Comment:
Is this reflective of on-chain state, or can a "force" mode be implemented?

Reply:
I'm for not force, because after force refund will be onchain representation in cache and if we have whales in cache then something goes wrong and let raise exception